OSSM-8268: Getting traffic into the mesh (#83901)
openshift/openshift-docs: modules/ossm-exposing-service-using-istio-gateway-and-virtualservice.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-exposing-service-using-istio-gateway-and-virtualservice_{context}"]

> = Exposing a service using an Istio gateway and a VirtualService

Reviewer: I think we called the procedure something like this in the content architecture call yesterday.

```suggestion
= Exposing a service using the Istio Gateway and VirtualService resources
```

Author: done

:context: ossm-exposing-service-using-istio-gateway-and-virtualservice

> This procedure configures a gateway that was deployed using gateway injection to expose a service in the mesh.

Reviewer: Adding some info here reinforcing that the gateway is configured using the Istio `Gateway` and `VirtualService` resources, and also making a distinction between exposing a service to traffic outside the mesh and exposing the gateway to traffic outside the cluster.

```suggestion
This procedure uses the Istio `Gateway` and `VirtualService` resources to configure a gateway that was deployed using gateway injection. The gateway is configured to expose a service within the mesh to traffic outside the mesh. After being configured, the gateway is exposed to traffic outside the cluster by setting the gateway `Service` to type `LoadBalancer`.
```

Author: chatted and updated.

.Prerequisites

* You have installed an {istio} gateway using gateway injection.

.Procedure

> . Deploy a service in mesh namespace. 

Reviewer: ```suggestion
. Deploy a sample service in a mesh namespace. 
```

Author: done

+
[source,terminal]
----
oc create namespace httpbin
oc label namespace httpbin istio-injection=enabled

> oc apply -n httpbin -f https://raw.githubusercontent.com/openshift-service-mesh/istio/refs/heads/master/samples/httpbin/httpbin.yaml

Reviewer: We probably need to lock the version down. Unfortunately, I don't think the version of httpbin in the latest `release-1.23` branch is working.

Hi @FilipB, do we know the name of the release branch that will be created for the `openshift-service-mesh/istio` repo for the TP2 release?

Reviewer 2: @eoinfennessy Sorry for delay. It's not yet clear. But I think this version from upstream istio 1.24 release branch should work too - https://raw.githubusercontent.com/istio/istio/refs/heads/release-1.24/samples/httpbin/httpbin.yaml

Author: done

. Create a `VirtualService` defining rules to route traffic from the gateway proxy to the sample service.
+
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: networking.istio.io/v1
kind: VirtualService
metadata:
  name: httpbin
  namespace: httpbin
spec:
  hosts:
  - httpbin.example.com <1>
  gateways:
  - httpbin-gateway <2>
  http:
  - match:
    - uri:
        prefix: /status
    - uri:
        prefix: /headers
    route:
    - destination: <3>
        port:
          number: 8000
        host: httpbin
----
<1> Specify the `hosts` that the routing rules of the `VirtualService` will be applied to. The `hosts` specified must be exposed by the {istio} `Gateway` resource the VirtualService is bound to.
<2> Bind the `VirtualService` to the {istio} `Gateway` resource created in the previous step by adding the `Gateway` name to the list of gateways.
<3> Route matching traffic to the sample service deployed earlier by defining a `destination` that includes the `host` and `port` of the httpbin `Service`.

. For verification purposes, deploy a `curl` client in a namespace outside of the mesh.
+
[source,terminal]
----
oc create namespace curl

> oc apply -n curl -f https://raw.githubusercontent.com/openshift-service-mesh/istio/refs/heads/master/samples/curl/curl.yaml

Reviewer: We also need to lock the version down here. The `curl` sample was only added to Istio recently, and is not present in the latest `release-1.23` branch.

Reviewer 2: I think we can also use upstream `release-1.24` branch here

Author: done.

----

. Using the `curl` client, send a request to the `/headers` endpoint of the httpbin application via the ingress gateway `Service`. Set the `Host` header of the request to `httpbin.example.com` to match the host specified in the {istio} `Gateway` and `VirtualService` resources.
+
[source,terminal]
----
CURL_POD=$(oc get pods -n curl -l app=curl -o jsonpath='{.items[*].metadata.name}')

oc exec $CURL_POD -n curl -- \
  curl -s -I \
    -H Host:httpbin.example.com \
    <gateway_name>.<gateway_namespace>.svc.cluster.local/headers
----

. The response should have a 200 OK HTTP status indicating that the request was successful.

> HTTP/1.1 200 OK
> server: istio-envoy

Reviewer: ```suggestion
.Example output
[source,terminal]
----
HTTP/1.1 200 OK
server: istio-envoy
...
----
```

Author: thanks for catching this! done.

. Send a curl request to an endpoint that does not have a corresponding URI prefix match defined in the httpbin `VirtualService`.
+
[source,terminal]
----
oc exec $CURL_POD -n curl -- \
  curl -s -I \
    -H Host:httpbin.example.com \
    <gateway_name>.<gateway_namespace>.svc.cluster.local/get
----
+
The response should return a `404 Not Found` status. This is expected because the `/get` endpoint does not have a matching URI prefix in the httpbin `VirtualService`.
+

> .Example output
> [source,terminal]
> ----
> HTTP/1.1 404 Not Found
> server: istio-envoy
> ----

Reviewer: Adding an ellipsis indicating that this is not the entire output.

```suggestion
.Example output
[source,terminal]
----
HTTP/1.1 404 Not Found
server: istio-envoy
...
----
```

Author: done

. Optional: Expose the gateway proxy to traffic outside the cluster by setting its `Service` type to `LoadBalancer`.
+
[source,terminal]
----
oc patch service <gateway_name> -n <gateway_namespace> -p '{"spec": {"type": "LoadBalancer"}}'
----
+
[NOTE]
====

> A gateway can also be exposed to traffic outside the cluster by using OpenShift Routes.

Reviewer: Something to this effect would be helpful. Maybe we could add a link?

```suggestion
A gateway can also be exposed to traffic outside the cluster by using OpenShift Routes. To learn how, see the "Expose a gateway to traffic outside the cluster using OpenShift Routes" procedure below.
```

Author: done

// This procedure is used in the following assembly:
// * gateways/ossm-getting-traffic-into-a-mesh.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-exposing-service-using-istio-gateway-and-virtualservice_{context}"]
= Exposing a service using the {istio} gateway and VirtualService resources
:context: ossm-exposing-service-using-istio-gateway-and-virtualservice

> This procedure uses the {istio} `Gateway` and `VirtualService` resources to configure a gateway that was deployed using gateway injection. The resources configure the gateway to expose a service in the mesh to traffic outside the mesh. Then, the gateway is exposed to traffic outside the cluster by setting the `Service` for the gateway to type `LoadBalancer`.

Author: done

.Prerequisites

* You have installed an {istio} gateway using gateway injection.

.Procedure

> . Deploy a sample service in a mesh namespace. 

Reviewer 3: May be we can rephrase this sentence to

`Deploy a sample service, named httpbin, in a mesh namespace.`

OR 

`Deploy an httpbin service in a mesh namespace.`

Author: done

+
[source,terminal]
----

> oc create namespace httpbin

Author: done

> oc label namespace httpbin istio-injection=enabled

Author: done

oc apply -n httpbin -f https://raw.githubusercontent.com/istio/istio/refs/heads/release-1.24/samples/httpbin/httpbin.yaml
----

. Create a YAML file named `httpbin-gw.yml` that defines an Istio `Gateway` resource. This resource configures gateway proxies to expose port 80 (HTTP) for the host, `httpbin.example.com`.
+
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: networking.istio.io/v1
kind: Gateway
metadata:
  name: httpbin-gateway
  namespace: httpbin
spec:
  selector:
    istio: <gateway_name> <1>
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:

>     - httpbin.example.com <2>

Author: done

----
<1> Set the `selector` to the unique label or set of labels specified in the pod template of the gateway proxy `Deployment`. By default, the {istio} `Gateway` resource configuration will apply to matching gateway pods in all namespaces.
<2> Using the `hosts` field, specify a list of addresses that can be used by clients when attempting to access a mesh service at the associated port.

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f httpbin-gw.yml
----

> . Create a YAML file named `httpbin-vs.yml` for a `VirtualService` The `VirtualService` defines the rules that route traffic from the gateway proxy to the sample service.

Reviewer 3: ```suggestion
. Create a YAML file named `httpbin-vs.yml` for a `VirtualService` The `VirtualService` defines the rules that route traffic from the gateway proxy to the httpbin service.
```

Author: done

+
[source,yaml,subs="attributes,verbatim"]
----
apiVersion: networking.istio.io/v1
kind: VirtualService
metadata:
  name: httpbin
  namespace: httpbin
spec:
  hosts:
  - httpbin.example.com <1>
  gateways:
  - httpbin-gateway <2>
  http:
  - match:
    - uri:
        prefix: /status
    - uri:
        prefix: /headers
    route:
    - destination: <3>
        port:
          number: 8000
        host: httpbin
----
<1> Specify the `hosts` that the routing rules of the `VirtualService` will be applied to. The `hosts` specified must be exposed by the {istio} `Gateway` resource the VirtualService is bound to.
<2> Bind the `VirtualService` to the {istio} `Gateway` resource created in the previous step by adding the `Gateway` name to the list of gateways.

> <3> Route matching traffic to the sample service deployed earlier by defining a `destination` that includes the `host` and `port` of the httpbin `Service`.

Reviewer 3: ```suggestion
<3> Route matching traffic to the sample `httpbin` service deployed earlier by defining a `destination` that includes the `host` and `port` of the httpbin `Service`.
```

Author: done

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f httpbin-vs.yml
----

. For verification purposes, deploy a `curl` client in a namespace outside of the mesh.
+
[source,terminal]
----
oc create namespace curl
oc apply -n curl -f https://raw.githubusercontent.com/istio/istio/refs/heads/release-1.24/samples/curl/curl.yaml
----

> . Using the `curl` client, send a request to the `/headers` endpoint of the httpbin application through the ingress gateway `Service`. Set the `Host` header of the request to `httpbin.example.com` to match the host that the {istio} `Gateway` and `VirtualService` resources specify.

Reviewer 4: ```suggestion
. Using the `curl` client, send a request to the `/headers` endpoint of the `httpbin` application through the ingress gateway `Service`. Set the `Host` header of the request to `httpbin.example.com` to match the host that the {istio} `Gateway` and `VirtualService` resources specify.
```

I see it varies in the core OCP documentation.

Author: done

+
[source,terminal]
----
CURL_POD=$(oc get pods -n curl -l app=curl -o jsonpath='{.items[*].metadata.name}')

oc exec $CURL_POD -n curl -- \
  curl -s -I \
    -H Host:httpbin.example.com \
    <gateway_name>.<gateway_namespace>.svc.cluster.local/headers
----

> . The response should have a 200 OK HTTP status indicating that the request was successful.

Reviewer 4: ```suggestion
. The response should have a `200 OK HTTP` status indicating that the request was successful.
```

Author: done

+
.Example output
[source,terminal]
----
HTTP/1.1 200 OK
server: istio-envoy
...
----

> . Send a curl request to an endpoint that does not have a corresponding URI prefix match defined in the httpbin `VirtualService`.

Reviewer 4: ```suggestion
. Send a curl request to an endpoint that does not have a corresponding URI prefix match defined in the httpbin `VirtualService`:
```

Author: done

  gateways:
  - httpbin-gateway <2>
  http:
  - match:
    - uri:
        prefix: /status
    - uri:
        prefix: /headers
    route:
    - destination: <3>
        port:
          number: 8000
        host: httpbin
# ...
----
<1> Specify the `hosts` that the routing rules of the `VirtualService` will be applied to. The `hosts` specified must be exposed by the {istio} `Gateway` resource the VirtualService is bound to.
<2> Bind the `VirtualService` to the {istio} `Gateway` resource created in the previous step by adding the `Gateway` name to the list of gateways.
<3> Route matching traffic to the `httpbin` service deployed earlier by defining a `destination` that includes the `host` and `port` of the httpbin `Service`.

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f httpbin-vs.yaml
----

. For verification purposes, create a namespace for a `curl` client:
+
[source,terminal]
----
$ oc create namespace curl
----

. Deploy the `curl` client by applying the curl.yaml file:
+
[source,terminal]
----
$ oc apply -n curl -f https://raw.githubusercontent.com/istio/istio/refs/heads/release-1.24/samples/curl/curl.yaml
----

> . Using the `curl` client, send a request to the `/headers` endpoint of the `httpbin` application through the ingress gateway `Service`. Set the `Host` header of the request to `httpbin.example.com` to match the host that the {istio} `Gateway` and `VirtualService` resources specify.

Reviewer 5: Need a resource, object, or CR. For follow-up.
```suggestion
. Using the `curl` client, send a request to the `/headers` endpoint of the `httpbin` application through the ingress gateway `Service` resource. Set the `Host` header of the request to `httpbin.example.com` to match the host that the {istio} `Gateway` and `VirtualService` resources specify.
```

Author: Captured as part of https://issues.redhat.com/browse/OSSM-8338

+
[source,terminal]
----
CURL_POD=$(oc get pods -n curl -l app=curl -o jsonpath='{.items[*].metadata.name}')

oc exec $CURL_POD -n curl -- \
  curl -s -I \
    -H Host:httpbin.example.com \
    <gateway_name>.<gateway_namespace>.svc.cluster.local/headers
----

. The response should have a `200 OK HTTP` status indicating that the request was successful.
+
.Example output
[source,terminal]
----
HTTP/1.1 200 OK
server: istio-envoy
...
----

. Send a curl request to an endpoint that does not have a corresponding URI prefix match defined in the httpbin `VirtualService`:
+
[source,terminal]
----
oc exec $CURL_POD -n curl -- \
  curl -s -I \
    -H Host:httpbin.example.com \
    <gateway_name>.<gateway_namespace>.svc.cluster.local/get
----
+

> The response should return a `404 Not Found` status. This is expected because the `/get` endpoint does not have a matching URI prefix in the httpbin `VirtualService`.

Reviewer 5: ```suggestion
The response should return a `404 Not Found` status. This is expected because the `/get` endpoint does not have a matching URI prefix in the `httpbin` `VirtualService` resource.
```

Author: captured as part of https://issues.redhat.com/browse/OSSM-8338
